driver_test.go

```diff
@@ -12,26 +12,246 @@
+	var err error
+
+	if emset{
+		adminClient, err = adminapi.NewDatabaseAdminClient(
+			ctx,
+			option.WithoutAuthentication(),
+			option.WithEndpoint(emhost),
+			option.WithGRPCDialOption(grpc.WithInsecure()))
+		if err != nil {
+			return nil, err
+		}
+	}else{
+		adminClient, err = adminapi.NewDatabaseAdminClient(ctx)
+		if err != nil {
+			return nil, err
+		}
 	}
-	connector, err := driver.OpenConnector("projects/$PROJECT/instances/$INSTANCE/databases/$DATABASE")
+
+	dataClient, err := spanner.NewClient(ctx, dsn)
 	if err != nil {
+		return nil, err
+	}
+
+	curs := &Connector{
+		ctx:         ctx,
+		client:      dataClient,
+		adminClient: adminClient,
+	}
+	return curs, nil
+}
+
+func (c *Connector) Close() {
+	c.client.Close()
+	c.adminClient.Close()
+}
+
+// structs for row data
+type testaRow struct {
+	A string
+	B string
+	C string
+}
+type typeTestaRow struct {
+	stringt string
+	intt    int
+	floatt  float64
+	boolt   bool
+}
+
+func init() {
+
+	// get environment variables or set to default
+	if instance, instset = os.LookupEnv("SPANNER_TEST_INSTANCE"); !instset {
+		instance = "test-instance"
+	}
+	if project, projset = os.LookupEnv("SPANNER_TEST_PROJECT"); !projset {
+		project = "test-project"
+	}
+	if dbid, dbset = os.LookupEnv("SPANNER_TEST_DBID"); !dbset {
+		dbid = "gotest"
+	}
+
+	// derive data source name
+	dsn = "projects/" + project + "/instances/" + instance + "/databases/" + dbid
+
+	// configure emulator if set 
+	emhost, emset = os.LookupEnv("SPANNER_EMULATOR_HOST")
+
+}
+
+// Executes DDL statements
+func executeDdlApi(curs *Connector, ddls []string) (err error) {
+
+	op, err := curs.adminClient.UpdateDatabaseDdl(curs.ctx, &adminpb.UpdateDatabaseDdlRequest{
+		Database:   dsn,
+		Statements: ddls,
+	})
+	if err != nil {
+		return err
+	}
+	if err := op.Wait(curs.ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// executes DML using the client library
+func ExecuteDMLClientLib(dml []string) (err error) {
+
+	// open client
+	ctx := context.Background()
+	client, err := spanner.NewClient(ctx, dsn)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	// Put strings into spanner.Statement structure
+	var states []spanner.Statement
+	for _, line := range dml {
+		states = append(states, spanner.NewStatement(line))
+	}
+
+	// execute statements
+	_, err = client.ReadWriteTransaction(ctx, func(ctx context.Context, txn *spanner.ReadWriteTransaction) error {
+		stmts := states
+		rowCounts, err := txn.BatchUpdate(ctx, stmts)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("Executed %d SQL statements using Batch DML.\n", len(rowCounts))
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Tests general query functionality
+func TestQueryContext(t *testing.T) {
+
+	// set up test table
+	curs, err := NewConnector(emhost,emset)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer curs.Close()
+
+	err = executeDdlApi(curs, []string{`CREATE TABLE Testa (
```

Comment:
I don't like this table name. Especially since there is a struct whose name depends on it. Maybe name the table based on the test i.e. `TestQueryContext`. That way the struct is `testQueryContextRow` which is much more descriptive.

Reply:
Changed to suggested names